File: src/WeaveFleet.Infrastructure/Plugins/BuiltIn/GitHub/GitHubItems.cs
using System.Globalization;
using System.Text;
using System.Text.Json.Nodes;

namespace WeaveFleet.Infrastructure.Plugins.BuiltIn.GitHub;

/// <summary>
/// Pull requests and issues as Fleet shows them in lists and on their pages, read from GitHub's GraphQL API:
/// one request carries a pull request's checks, review decision and reviewers, which the REST API spreads
/// over several calls per pull request.
/// </summary>
internal static class GitHubItems
{
    private const string PullRequestFields = """
        fragment Pr on PullRequest {
          __typename number title url state isDraft createdAt updatedAt
          author { login avatarUrl }
          repository { name owner { login } }
          comments { totalCount }
          labels(first: 10) { nodes { name color } }
          headRefName baseRefName additions deletions changedFiles mergeable reviewDecision
          latestOpinionatedReviews(first: 10) { nodes { state author { login avatarUrl } } }
          reviewRequests(first: 10) { nodes { requestedReviewer { __typename ... on User { login avatarUrl } ... on Team { name } } } }
          commits(last: 1) { nodes { commit { statusCheckRollup { state } } } }
        }
        """;

    private const string IssueFields = """
        fragment Iss on Issue {
          __typename number title url state createdAt updatedAt
          author { login avatarUrl }
          repository { name owner { login } }
          comments { totalCount }
          labels(first: 10) { nodes { name color } }
          assignees(first: 5) { nodes { login } }
        }
        """;

    internal const string SearchQuery = PullRequestFields + IssueFields + """
        query($q: String!, $first: Int!, $after: String) {
          search(query: $q, type: ISSUE, first: $first, after: $after) {
            issueCount
            pageInfo { endCursor hasNextPage }
            nodes { ...Pr ...Iss }
          }
        }
        """;

    internal const string WorkQuery = PullRequestFields + IssueFields + """
        query($review: String!, $authored: String!, $assigned: String!) {
          viewer { login }
          review: search(query: $review, type: ISSUE, first: 20) { nodes { ...Pr } }
          authored: search(query: $authored, type: ISSUE, first: 20) { nodes { ...Pr } }
          assigned: search(query: $assigned, type: ISSUE, first: 20) { nodes { ...Iss ...Pr } }
        }
        """;

    internal const string DetailQuery = PullRequestFields + IssueFields + """
        query($owner: String!, $repo: String!, $number: Int!) {
          repository(owner: $owner, name: $repo) {
            issueOrPullRequest(number: $number) {
              ...Pr
              ...Iss
              ... on PullRequest {
                body mergedAt
                mergedBy { login }
                reviewThreads(first: 100) { nodes { isResolved isOutdated } }
                commits(last: 1) {
                  nodes {
                    commit {
                      statusCheckRollup {
                        contexts(first: 100) {
                          nodes {
                            __typename
                            ... on CheckRun {
                              name status conclusion detailsUrl startedAt completedAt databaseId
                              checkSuite { workflowRun { workflow { name } } }
                            }
                            ... on StatusContext { context state targetUrl createdAt }
                          }
                        }
                      }
                    }
                  }
                }
                timelineItems(last: 60, itemTypes: [ISSUE_COMMENT, PULL_REQUEST_REVIEW, MERGED_EVENT, CLOSED_EVENT, REOPENED_EVENT, READY_FOR_REVIEW_EVENT, CONVERT_TO_DRAFT_EVENT, CROSS_REFERENCED_EVENT]) {
                  nodes { ...Timeline }
                }
              }
              ... on Issue {
                body
                timelineItems(last: 60, itemTypes: [ISSUE_COMMENT, CLOSED_EVENT, REOPENED_EVENT, CROSS_REFERENCED_EVENT]) {
                  nodes { ...Timeline }
                }
              }
            }
          }
        }

        fragment Timeline on Node {
          __typename
          ... on IssueComment { author { login avatarUrl } createdAt body url }
          ... on PullRequestReview { author { login avatarUrl } submittedAt createdAt body state url }
          ... on MergedEvent { actor { login avatarUrl } createdAt }
          ... on ClosedEvent { actor { login avatarUrl } createdAt }
          ... on ReopenedEvent { actor { login avatarUrl } createdAt }
          ... on ReadyForReviewEvent { actor { login avatarUrl } createdAt }
          ... on ConvertToDraftEvent { actor { login avatarUrl } createdAt }
          ... on CrossReferencedEvent {
            actor { login avatarUrl } createdAt
            source {
              __typename
              ... on PullRequest { number title url state repository { nameWithOwner } }
              ... on Issue { number title url state repository { nameWithOwner } }
            }
          }
        }
        """;

    /// <summary>
    /// A query for items picked by number. GraphQL has no "these numbers" filter, so each number gets its own
    /// aliased field (<c>i0</c>, <c>i1</c>…) in one request.
    /// </summary>
    internal static (string Query, JsonObject Variables) BuildNumbersQuery(string owner, string repo, IReadOnlyList<int> numbers)
    {
        var fields = new StringBuilder();
        for (var i = 0; i < numbers.Count; i++)
        {
            fields.Append(CultureInfo.InvariantCulture, $"i{i}: issueOrPullRequest(number: {numbers[i]}) {{ ...Pr ...Iss }}\n");
        }

        var query = PullRequestFields + IssueFields
            + "query($owner: String!, $repo: String!) {\n  repository(owner: $owner, name: $repo) {\n"
            + fields
            + "  }\n}\n";
        return (query, new JsonObject { ["owner"] = owner, ["repo"] = repo });
    }

    /// <summary>Open pull request and issue counts for each repository, one aliased field per repository.</summary>
    internal static (string Query, JsonObject Variables) BuildRepoCountsQuery(IReadOnlyList<(string Owner, string Name)> repos)
    {
        var parameters = new StringBuilder();
        var fields = new StringBuilder();
        var variables = new JsonObject();
        for (var i = 0; i < repos.Count; i++)
        {
            parameters.Append(CultureInfo.InvariantCulture, $"$o{i}: String!, $n{i}: String!, ");
            fields.Append(CultureInfo.InvariantCulture,
                $"r{i}: repository(owner: $o{i}, name: $n{i}) {{ nameWithOwner pullRequests(states: OPEN) {{ totalCount }} issues(states: OPEN) {{ totalCount }} }}\n");
            variables[$"o{i}"] = repos[i].Owner;
            variables[$"n{i}"] = repos[i].Name;
        }

        return ($"query({parameters.ToString().TrimEnd(' ', ',')}) {{\n{fields}}}\n", variables);
    }

    /// <summary>The message from GitHub's first GraphQL error, when the response carries one.</summary>
    internal static string? ErrorMessage(JsonNode? response)
        => (response?["errors"] as JsonArray)?.OfType<JsonObject>().Select(e => e["message"]?.GetValue<string>()).FirstOrDefault(m => m is not null);

    internal static GitHubItemPage ParseSearch(JsonNode? response)
    {
        var search = response?["data"]?["search"];
        return new GitHubItemPage(
            ParseNodes(search?["nodes"] as JsonArray),
            search?["issueCount"]?.GetValue<int>() ?? 0,
            search?["pageInfo"]?["endCursor"]?.GetValue<string>(),
            search?["pageInfo"]?["hasNextPage"]?.GetValue<bool>() ?? false);
    }

    internal static GitHubWorkResponse ParseWork(JsonNode? response, IReadOnlyList<GitHubRepoCounts> repos)
    {
        var data = response?["data"];
        return new GitHubWorkResponse(
            data?["viewer"]?["login"]?.GetValue<string>(),
            ParseNodes(data?["review"]?["nodes"] as JsonArray),
            ParseNodes(data?["authored"]?["nodes"] as JsonArray),
            ParseNodes(data?["assigned"]?["nodes"] as JsonArray),
            repos);
    }

    internal static IReadOnlyList<GitHubRepoCounts> ParseRepoCounts(JsonNode? response)
    {
        if (response?["data"] is not JsonObject data)
            return [];

        return data
            .Select(pair => pair.Value)
            .OfType<JsonObject>()
            .Select(repo => new GitHubRepoCounts(
                repo["nameWithOwner"]?.GetValue<string>() ?? string.Empty,
                repo["pullRequests"]?["totalCount"]?.GetValue<int>() ?? 0,
                repo["issues"]?["totalCount"]?.GetValue<int>() ?? 0))
            .Where(repo => repo.FullName.Length > 0)
            .ToArray();
    }

    /// <summary>Items picked by number, in the order asked for; numbers GitHub doesn't know are left out.</summary>
    internal static IReadOnlyList<GitHubItemSummary> ParseNumbers(JsonNode? response, int count)
    {
        var repository = response?["data"]?["repository"];
        var items = new List<GitHubItemSummary>();
        for (var i = 0; i < count; i++)
        {
            if (repository?[$"i{i}"] is JsonObject node && ParseSummary(node) is { } item)
                items.Add(item);
        }

        return items;
    }

    internal static GitHubItemDetail? ParseDetail(JsonNode? response)
    {
        if (response?["data"]?["repository"]?["issueOrPullRequest"] is not JsonObject node || ParseSummary(node) is not { } summary)
            return null;

        var threads = node["reviewThreads"]?["nodes"] as JsonArray ?? [];
        var contexts = node["commits"]?["nodes"]?[0]?["commit"]?["statusCheckRollup"]?["contexts"]?["nodes"] as JsonArray ?? [];

        return new GitHubItemDetail(
            summary,
            node["body"]?.GetValue<string>(),
            node["changedFiles"]?.GetValue<int>(),
            node["mergedAt"]?.GetValue<string>(),
            node["mergedBy"]?["login"]?.GetValue<string>(),
            contexts.OfType<JsonObject>().Select(ParseCheck).OfType<GitHubCheckDto>().ToArray(),
            threads.OfType<JsonObject>().Count(t => t["isResolved"]?.GetValue<bool>() != true && t["isOutdated"]?.GetValue<bool>() != true),
            (node["timelineItems"]?["nodes"] as JsonArray ?? []).OfType<JsonObject>().Select(ParseTimelineEntry).OfType<GitHubTimelineEntry>().ToArray());
    }

    private static GitHubItemSummary[] ParseNodes(JsonArray? nodes)
        => (nodes ?? []).OfType<JsonObject>().Select(ParseSummary).OfType<GitHubItemSummary>().ToArray();

    internal static GitHubItemSummary? ParseSummary(JsonObject node)
    {
        var typeName = node["__typename"]?.GetValue<string>();
        if (typeName is not ("PullRequest" or "Issue"))
            return null;

        var isPull = typeName == "PullRequest";
        var labels = (node["labels"]?["nodes"] as JsonArray ?? [])
            .OfType<JsonObject>()
            .Select(l => new GitHubLabelDto(l["name"]?.GetValue<string>() ?? string.Empty, l["color"]?.GetValue<string>() ?? string.Empty))
            .ToArray();

        return new GitHubItemSummary(
            isPull ? "pull" : "issue",
            node["repository"]?["owner"]?["login"]?.GetValue<string>() ?? string.Empty,
            node["repository"]?["name"]?.GetValue<string>() ?? string.Empty,
            node["number"]?.GetValue<int>() ?? 0,
            node["title"]?.GetValue<string>() ?? string.Empty,
            node["url"]?.GetValue<string>() ?? string.Empty,
            (node["state"]?.GetValue<string>() ?? "OPEN").ToLowerInvariant(),
            node["isDraft"]?.GetValue<bool>() == true,
            node["author"]?["login"]?.GetValue<string>(),
            node["author"]?["avatarUrl"]?.GetValue<string>(),
            node["createdAt"]?.GetValue<string>() ?? string.Empty,
            node["updatedAt"]?.GetValue<string>() ?? string.Empty,
            node["comments"]?["totalCount"]?.GetValue<int>() ?? 0,
            labels,
            isPull ? node["headRefName"]?.GetValue<string>() : null,
            isPull ? node["baseRefName"]?.GetValue<string>() : null,
            isPull ? node["additions"]?.GetValue<int>() : null,
            isPull ? node["deletions"]?.GetValue<int>() : null,
            isPull ? ChecksRollup(node["commits"]?["nodes"]?[0]?["commit"]?["statusCheckRollup"]?["state"]?.GetValue<string>()) : null,
            isPull ? node["reviewDecision"]?.GetValue<string>() : null,
            isPull ? node["mergeable"]?.GetValue<string>() : null,
            isPull ? ParseReviewers(node) : [],
            (node["assignees"]?["nodes"] as JsonArray ?? []).OfType<JsonObject>()
                .Select(a => a["login"]?.GetValue<string>()).OfType<string>().ToArray());
    }

    /// <summary>"success", "failure", "pending" or "none", from GitHub's rollup of a commit's checks.</summary>
    internal static string ChecksRollup(string? rollupState) => rollupState switch
    {
        "SUCCESS" => "success",
        "FAILURE" or "ERROR" => "failure",
        "PENDING" or "EXPECTED" => "pending",
        _ => "none",
    };

    /// <summary>
    /// Everyone with a say on a pull request: each reviewer's latest verdict, then anyone asked who hasn't
    /// answered yet (state <c>REQUESTED</c>).
    /// </summary>
    internal static GitHubReviewerDto[] ParseReviewers(JsonNode pullRequest)
    {
        var reviewers = new List<GitHubReviewerDto>();
        foreach (var review in (pullRequest["latestOpinionatedReviews"]?["nodes"] as JsonArray ?? []).OfType<JsonObject>())
        {
            var login = review["author"]?["login"]?.GetValue<string>();
            if (login is null || reviewers.Any(r => r.Login == login))
                continue;
            reviewers.Add(new GitHubReviewerDto(login, review["author"]?["avatarUrl"]?.GetValue<string>(), review["state"]?.GetValue<string>() ?? "COMMENTED"));
        }

        foreach (var request in (pullRequest["reviewRequests"]?["nodes"] as JsonArray ?? []).OfType<JsonObject>())
        {
            var reviewer = request["requestedReviewer"];
            var login = reviewer?["login"]?.GetValue<string>() ?? reviewer?["name"]?.GetValue<string>();
            if (login is null || reviewers.Any(r => r.Login == login))
                continue;
            reviewers.Add(new GitHubReviewerDto(login, reviewer?["avatarUrl"]?.GetValue<string>(), "REQUESTED"));
        }

        return reviewers.ToArray();
    }

    private static GitHubCheckDto? ParseCheck(JsonObject context)
    {
        if (context["__typename"]?.GetValue<string>() == "StatusContext")
        {
            var state = context["state"]?.GetValue<string>();
            return new GitHubCheckDto(
                context["context"]?.GetValue<string>() ?? string.Empty,
                state switch { "SUCCESS" => "success", "FAILURE" or "ERROR" => "failure", _ => "pending" },
                context["targetUrl"]?.GetValue<string>(),
                null,
                null,
                context["createdAt"]?.GetValue<string>(),
                null);
        }

        if (context["__typename"]?.GetValue<string>() != "CheckRun")
            return null;

        var status = context["status"]?.GetValue<string>();
        var conclusion = context["conclusion"]?.GetValue<string>();
        var checkState = status != "COMPLETED"
            ? "pending"
            : conclusion switch
            {
                "SUCCESS" => "success",
                "FAILURE" or "TIMED_OUT" or "STARTUP_FAILURE" or "ACTION_REQUIRED" => "failure",
                "SKIPPED" => "skipped",
                _ => "neutral",
            };

        return new GitHubCheckDto(
            context["name"]?.GetValue<string>() ?? string.Empty,
            checkState,
            context["detailsUrl"]?.GetValue<string>(),
            context["checkSuite"]?["workflowRun"]?["workflow"]?["name"]?.GetValue<string>(),
            context["databaseId"]?.GetValue<long>(),
            context["startedAt"]?.GetValue<string>(),
            context["completedAt"]?.GetValue<string>());
    }

    private static GitHubTimelineEntry? ParseTimelineEntry(JsonObject item)
    {
        var who = item["author"] ?? item["actor"];
        var login = who?["login"]?.GetValue<string>();
        var avatar = who?["avatarUrl"]?.GetValue<string>();
        var at = item["submittedAt"]?.GetValue<string>() ?? item["createdAt"]?.GetValue<string>() ?? string.Empty;

        switch (item["__typename"]?.GetValue<string>())
        {
            case "IssueComment":
                return new GitHubTimelineEntry("comment", login, avatar, at, item["body"]?.GetValue<string>(), null, item["url"]?.GetValue<string>(), null);
            case "PullRequestReview":
                var state = item["state"]?.GetValue<string>();
                var body = item["body"]?.GetValue<string>();
                // A bare "commented" review is the wrapper around inline comments; the threads show those.
                if (state == "COMMENTED" && string.IsNullOrWhiteSpace(body))
                    return null;
                return new GitHubTimelineEntry("review", login, avatar, at, body, state, item["url"]?.GetValue<string>(), null);
            case "MergedEvent":
                return new GitHubTimelineEntry("merged", login, avatar, at, null, null, null, null);
            case "ClosedEvent":
                return new GitHubTimelineEntry("closed", login, avatar, at, null, null, null, null);
            case "ReopenedEvent":
                return new GitHubTimelineEntry("reopened", login, avatar, at, null, null, null, null);
            case "ReadyForReviewEvent":
                return new GitHubTimelineEntry("ready", login, avatar, at, null, null, null, null);
            case "ConvertToDraftEvent":
                return new GitHubTimelineEntry("draft", login, avatar, at, null, null, null, null);
            case "CrossReferencedEvent":
                var source = item["source"];
                var number = source?["number"]?.GetValue<int>();
                if (number is null)
                    return null;
                var kind = source?["__typename"]?.GetValue<string>() == "PullRequest" ? "pull" : "issue";
                return new GitHubTimelineEntry(
                    "referenced",
                    login,
                    avatar,
                    at,
                    source?["title"]?.GetValue<string>(),
                    source?["state"]?.GetValue<string>()?.ToLowerInvariant(),
                    source?["url"]?.GetValue<string>(),
                    new GitHubReferenceDto(kind, source?["repository"]?["nameWithOwner"]?.GetValue<string>() ?? string.Empty, number.Value));
            default:
                return null;
        }
    }
}

// ── Response types (file-level so ApiJsonContext in the Api project can see them) ─────────────────

/// <summary>A label and its hex colour, without the <c>#</c>.</summary>
public sealed record GitHubLabelDto(string Name, string Color);

/// <summary>
/// Someone with a say on a pull request. <c>State</c> is their latest verdict (APPROVED, CHANGES_REQUESTED,
/// COMMENTED, DISMISSED), or REQUESTED while they haven't answered.
/// </summary>
public sealed record GitHubReviewerDto(string Login, string? AvatarUrl, string State);

/// <summary>A check on a pull request's head commit. <c>State</c>: success, failure, pending, skipped or neutral.</summary>
public sealed record GitHubCheckDto(
    string Name,
    string State,
    string? Url,
    string? WorkflowName,
    long? CheckRunId,
    string? StartedAt,
    string? CompletedAt);

/// <summary>
/// A pull request or issue as a list row shows it. The pull request fields are null for issues.
/// <c>State</c>: open, closed or merged. <c>Checks</c>: success, failure, pending or none.
/// <c>ReviewDecision</c>: APPROVED, CHANGES_REQUESTED, REVIEW_REQUIRED or null.
/// <c>Mergeable</c>: MERGEABLE, CONFLICTING or UNKNOWN.
/// </summary>
public sealed record GitHubItemSummary(
    string Kind,
    string Owner,
    string Repo,
    int Number,
    string Title,
    string Url,
    string State,
    bool IsDraft,
    string? Author,
    string? AuthorAvatarUrl,
    string CreatedAt,
    string UpdatedAt,
    int Comments,
    IReadOnlyList<GitHubLabelDto> Labels,
    string? HeadRef,
    string? BaseRef,
    int? Additions,
    int? Deletions,
    string? Checks,
    string? ReviewDecision,
    string? Mergeable,
    IReadOnlyList<GitHubReviewerDto> Reviewers,
    IReadOnlyList<string> Assignees);

/// <summary>One page of search results.</summary>
public sealed record GitHubItemPage(IReadOnlyList<GitHubItemSummary> Items, int TotalCount, string? EndCursor, bool HasNextPage);

/// <summary>Another pull request or issue, as a timeline entry points at it.</summary>
public sealed record GitHubReferenceDto(string Kind, string Repository, int Number);

/// <summary>
/// Something that happened on a pull request or issue. <c>Kind</c>: comment, review, merged, closed, reopened,
/// ready, draft or referenced. A review's <c>State</c> is its verdict; a reference's is the other item's state.
/// </summary>
public sealed record GitHubTimelineEntry(
    string Kind,
    string? Author,
    string? AuthorAvatarUrl,
    string CreatedAt,
    string? Body,
    string? State,
    string? Url,
    GitHubReferenceDto? Reference);

/// <summary>What a pull request or issue page shows. The pull request fields are null or empty for issues.</summary>
public sealed record GitHubItemDetail(
    GitHubItemSummary Summary,
    string? Body,
    int? ChangedFiles,
    string? MergedAt,
    string? MergedBy,
    IReadOnlyList<GitHubCheckDto> Checks,
    int UnresolvedThreads,
    IReadOnlyList<GitHubTimelineEntry> Timeline);

/// <summary>Open pull request and issue counts for a followed repository.</summary>
public sealed record GitHubRepoCounts(string FullName, int OpenPullRequests, int OpenIssues);

/// <summary>
/// The GitHub home: pull requests waiting on the user's review, the user's own open pull requests, and
/// issues assigned to them, across the repositories they follow.
/// </summary>
public sealed record GitHubWorkResponse(
    string? Login,
    IReadOnlyList<GitHubItemSummary> ReviewRequested,
    IReadOnlyList<GitHubItemSummary> Authored,
    IReadOnlyList<GitHubItemSummary> Assigned,
    IReadOnlyList<GitHubRepoCounts> Repos);

File: client/src/lib/github-items.ts
/**
 * Pull requests and issues as the GitHub pages show them: list rows (`GitHubItemSummary`) and a page
 * (`GitHubItemDetail`), from Fleet's GitHub endpoints, which read them from GitHub's GraphQL API.
 */
import { apiFetch } from "@/lib/api-client"
import { createGitHubSessionSourcePreset, type GitHubSessionSourcePreset } from "@/lib/github-session-source"
import type { IssueFilterState } from "@/plugins/builtin/github/composables/github-types"
import { checksFromCounts, checksFromRollup, parseReviewers, prState, prWords, reviewDecision, type CheckCounts, type PrFacts, type Reviewer } from "@/lib/pr-state"
import { linkDiff, linkLabels, linkNumber, linkReviewers, linkTitle, summarizeChecks, isPullRequest, linkHref, type SmartLink } from "@/lib/smart-links"

export interface GitHubItemLabel {
  name: string
  color: string
}

export interface GitHubItemSummary {
  kind: "pull" | "issue"
  owner: string
  repo: string
  number: number
  title: string
  url: string
  /** open, closed or merged. */
  state: string
  isDraft: boolean
  author: string | null
  authorAvatarUrl: string | null
  createdAt: string
  updatedAt: string
  comments: number
  labels: GitHubItemLabel[]
  headRef: string | null
  baseRef: string | null
  additions: number | null
  deletions: number | null
  /** success, failure, pending or none; null for issues. */
  checks: string | null
  reviewDecision: string | null
  /** MERGEABLE, CONFLICTING or UNKNOWN. */
  mergeable: string | null
  reviewers: Reviewer[]
  assignees: string[]
}

export interface GitHubItemPage {
  items: GitHubItemSummary[]
  totalCount: number
  endCursor: string | null
  hasNextPage: boolean
}

export interface GitHubCheck {
  name: string
  /** success, failure, pending, skipped or neutral. */
  state: "success" | "failure" | "pending" | "skipped" | "neutral"
  url: string | null
  workflowName: string | null
  checkRunId: number | null
  startedAt: string | null
  completedAt: string | null
}

export interface GitHubTimelineEntry {
  kind: "comment" | "review" | "merged" | "closed" | "reopened" | "ready" | "draft" | "referenced"
  author: string | null
  authorAvatarUrl: string | null
  createdAt: string
  body: string | null
  /** A review's verdict, or the referencing item's state. */
  state: string | null
  url: string | null
  reference: { kind: "pull" | "issue"; repository: string; number: number } | null
}

export interface GitHubItemDetail {
  summary: GitHubItemSummary
  body: string | null
  changedFiles: number | null
  mergedAt: string | null
  mergedBy: string | null
  checks: GitHubCheck[]
  unresolvedThreads: number
  timeline: GitHubTimelineEntry[]
}

export interface GitHubRepoCounts {
  fullName: string
  openPullRequests: number
  openIssues: number
}

export interface GitHubWork {
  login: string | null
  avatarUrl: string | null
  reviewRequested: GitHubItemSummary[]
  authored: GitHubItemSummary[]
  assigned: GitHubItemSummary[]
  repos: GitHubRepoCounts[]
}

/** A GitHub request that failed, with the message to show. */
export class GitHubItemsError extends Error {
  constructor(message: string, readonly status: number) {
    super(message)
  }
}

async function getJson<T>(path: string, signal?: AbortSignal): Promise<T> {
  const response = await apiFetch(path, { signal })
  if (!response.ok) {
    const payload = (await response.json().catch(() => ({}))) as { error?: string }
    const message = response.status === 401
      ? "Connect GitHub in Settings to see this."
      : payload.error ?? `GitHub didn't answer (HTTP ${response.status}).`
    throw new GitHubItemsError(message, response.status)
  }
  return normalize(await response.json()) as T
}

// Reviewers arrive as plain objects; give them the typed shape everywhere they appear.
function normalize(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(normalize)
  if (!value || typeof value !== "object") return value
  const entries = Object.entries(value as Record<string, unknown>).map(([key, entry]) =>
    [key, key === "reviewers" ? parseReviewers(entry) : normalize(entry)])
  return Object.fromEntries(entries)
}

const encode = encodeURIComponent

export function searchGitHubItems(q: string, options: { after?: string | null; first?: number; signal?: AbortSignal } = {}): Promise<GitHubItemPage> {
  const params = new URLSearchParams({ q, first: String(options.first ?? 30) })
  if (options.after) params.set("after", options.after)
  return getJson(`/api/integrations/github/search?${params.toString()}`, options.signal)
}

export function fetchGitHubItemsByNumber(owner: string, repo: string, numbers: number[], signal?: AbortSignal): Promise<GitHubItemSummary[]> {
  if (numbers.length === 0) return Promise.resolve([])
  return getJson(`/api/integrations/github/repos/${encode(owner)}/${encode(repo)}/items?numbers=${numbers.join(",")}`, signal)
}

export function fetchGitHubItem(owner: string, repo: string, number: number, signal?: AbortSignal): Promise<GitHubItemDetail> {
  return getJson(`/api/integrations/github/repos/${encode(owner)}/${encode(repo)}/items/${number}`, signal)
}

export function fetchGitHubWork(signal?: AbortSignal): Promise<GitHubWork> {
  return getJson("/api/integrations/github/work", signal)
}

/** A body as GitHub shows it: HTML comments (bots leave notes in them) are hidden there, so they are here. */
export function visibleBody(body: string | null | undefined): string {
  return (body ?? "").replace(/<!--[\s\S]*?-->/g, "").trim()
}

/** `owner/repo#123`, the key smart links use. */
export const itemResourceId = (item: Pick<GitHubItemSummary, "owner" | "repo" | "number">): string =>
  `${item.owner}/${item.repo}#${item.number}`

export function checkCounts(checks: GitHubCheck[]): CheckCounts {
  return {
    passed: checks.filter((c) => c.state === "success" || c.state === "skipped" || c.state === "neutral").length,
    failing: checks.filter((c) => c.state === "failure").length,
    pending: checks.filter((c) => c.state === "pending").length,
  }
}

/** What the pull request state language needs from a list row, or from a page when it has the checks. */
export function itemPrFacts(item: GitHubItemSummary, detail?: Pick<GitHubItemDetail, "checks" | "unresolvedThreads">): PrFacts {
  const counts = detail && detail.checks.length > 0 ? checkCounts(detail.checks) : undefined
  return {
    state: item.state === "merged" ? "merged" : item.state === "closed" ? "closed" : "open",
    draft: item.isDraft,
    checks: counts ? checksFromCounts(counts) : checksFromRollup(item.checks),
    conflict: item.mergeable === "CONFLICTING",
    review: reviewDecision(item.reviewDecision),
    unresolvedThreads: detail?.unresolvedThreads ?? 0,
    checkCounts: counts,
  }
}

/** What "Start session" hands the new-session composer. */
export function itemSessionPreset(item: GitHubItemSummary, body: string | null = null): GitHubSessionSourcePreset {
  const isPull = item.kind === "pull"
  return createGitHubSessionSourcePreset({
    sourceType: isPull ? "github-pull-request" : "github-issue",
    owner: item.owner,
    repo: item.repo,
    number: item.number,
    title: item.title,
    body,
    htmlUrl: item.url,
    repoFullName: `${item.owner}/${item.repo}`,
    suggestedBranch: isPull ? item.headRef : null,
  })
}

const quote = (value: string) => (/\s/.test(value) ? `"${value}"` : value)

/** The issue filter bar's choices as GitHub search qualifiers, e.g. `is:open label:bug sort:updated-desc`. */
export function issueFilterQuery(filter: IssueFilterState): string {
  const parts: string[] = []
  if (filter.state !== "all") parts.push(`is:${filter.state}`)
  for (const label of filter.labels) parts.push(`label:${quote(label)}`)
  if (filter.author) parts.push(`author:${filter.author}`)
  if (filter.assignee === "none") parts.push("no:assignee")
  else if (filter.assignee && filter.assignee !== "*") parts.push(`assignee:${filter.assignee}`)
  if (filter.milestone) parts.push(`milestone:${quote(filter.milestone)}`)
  if (filter.type) parts.push(`type:${quote(filter.type)}`)
  parts.push(`sort:${filter.sort}-${filter.direction}`)
  if (filter.search.trim()) parts.push(filter.search.trim())
  return parts.join(" ")
}

/**
 * A session's link as a list row, for "In Fleet": the watcher already keeps its state, checks and reviews,
 * so the row needs no request to GitHub.
 */
export function summaryFromLink(link: SmartLink): GitHubItemSummary | null {
  const number = linkNumber(link)
  const { owner, repo } = link.metadata
  if (number === null || typeof owner !== "string" || typeof repo !== "string") return null
  const isPull = isPullRequest(link)
  const checks = summarizeChecks(link)
  const diff = linkDiff(link)
  const meta = (key: string) => (typeof link.metadata[key] === "string" ? (link.metadata[key] as string) : null)
  return {
    kind: isPull ? "pull" : "issue",
    owner,
    repo,
    number,
    title: linkTitle(link),
    url: linkHref(link),
    state: link.status === "draft" ? "open" : link.status || "open",
    isDraft: link.status === "draft",
    author: meta("author"),
    authorAvatarUrl: meta("authorAvatarUrl"),
    createdAt: link.createdAt,
    // When GitHub last saw a change, not when Fleet last checked.
    updatedAt: meta("updatedAt") ?? link.updatedAt,
    comments: 0,
    labels: linkLabels(link),
    headRef: meta("headRef"),
    baseRef: meta("baseRef"),
    additions: diff?.additions ?? null,
    deletions: diff?.deletions ?? null,
    checks: isPull ? (checks.state === "failing" ? "failure" : checks.state === "running" ? "pending" : checks.state === "passed" ? "success" : "none") : null,
    reviewDecision: meta("reviewDecision"),
    mergeable: link.metadata.mergeable === false ? "CONFLICTING" : link.metadata.mergeable === true ? "MERGEABLE" : null,
    reviewers: linkReviewers(link),
    assignees: [],
  }
}

export interface NeedsYouItem {
  item: GitHubItemSummary
  /** Why it's here: "Review requested", "Checks failing", "Changes requested". */
  reason: string
}

/**
 * What needs the user: pull requests waiting on their review, then their own pull requests something blocks
 * (a failing check, a conflict, requested changes).
 */
export function needsYou(work: GitHubWork): NeedsYouItem[] {
  const blocked = work.authored
    .filter((item) => prState(itemPrFacts(item)) === "blocked")
    .map((item) => ({ item, reason: prWords(itemPrFacts(item)) }))
  return [...work.reviewRequested.map((item) => ({ item, reason: "Review requested" })), ...blocked]
}

/** The user's open pull requests that nothing blocks. */
export function yourPullRequests(work: GitHubWork): GitHubItemSummary[] {
  return work.authored.filter((item) => prState(itemPrFacts(item)) !== "blocked")
}

/** The Fleet route for a pull request or issue page. */
export function itemRoute(item: Pick<GitHubItemSummary, "kind" | "owner" | "repo" | "number">): string {
  return `/github/${encodeURIComponent(item.owner)}/${encodeURIComponent(item.repo)}/${item.kind === "pull" ? "pulls" : "issues"}/${item.number}`
}
